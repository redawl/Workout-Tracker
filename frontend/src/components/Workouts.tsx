import { 
    Title, 
    Panel, 
    PanelHeader, 
    Divider, 
    PanelMain, 
    PanelMainBody, 
    DatePicker, 
    Form, 
    FormGroup,
    ActionGroup,
    Button,
    ButtonVariant,
    TextInput,
    NumberInput,
    Alert
} from '@patternfly/react-core';
import { REP_STEP, SET_STEP, WEIGHT_STEP, WORKOUTS } from './constants';
import React from 'react';
import { formatDate } from '../helpers/DateHelpers';
import {     
    Thead,
    Tr,
    Th,
    Table,
    Tbody,
    Td
} from '@patternfly/react-table'

import { WorkoutService } from '../workout-api-client/services.gen';
import { Workout, Exercise } from '../workout-api-client/types.gen';
import { ApiError } from '../workout-api-client';

export const Workouts = () => {
    const [workoutDate, setWorkoutDate] = React.useState(formatDate(new Date()));
    const [exercises, setExercises] = React.useState([] as Array<Exercise>);
    const [saveSuccess, setSaveSuccess] = React.useState(false);
    const [saveFailureMessage, setSaveFailureMessage] = React.useState('');
    const [unsavedChanges, setUnsavedChanges] = React.useState(false);

    const workoutExists = React.useRef(false);

    React.useEffect(() => {
        WorkoutService.getWorkoutByDate({ date: workoutDate })
            .then((workout: Workout) => {
                    workoutExists.current = true;
                setExercises(workout.exercises as Array<Exercise>);
                setUnsavedChanges(false);
            })
            .catch((apiError: ApiError) => {
                if(apiError.status === 404){
                    workoutExists.current = false;
                    setExercises([]);
                } else {
                    console.error(apiError);
                }
            });
    }, [workoutDate]);

    const newExercise = () => {
        setExercises([...exercises,
            {
                name: "",
                lbs: 0,
                sets: 0,
                reps: 0
            }
        ])
    }

    const removeExercise = (exerciseIndex: number) => {
        setExercises(exercises.filter((_exercise, index) => index !== exerciseIndex));
    }

    const updateName = (exerciseIndex: number, exerciseName: string) => {
        setExercises(exercises.map((exercise, index) => {
            if(index === exerciseIndex) {
                setUnsavedChanges(true);
                return { ...exercise, name: exerciseName };
            }
            return exercise;
        }))
    }

    const updateLbs = (exerciseIndex: number, lbs: number) => {
        setExercises(exercises.map((exercise, index) => {
            if(index === exerciseIndex) {
                setUnsavedChanges(true);
                return { ...exercise, lbs };
            }

            return exercise;
        }));
    } 

    const updateSets = (exerciseIndex: number, sets: number) => {
        setExercises(exercises.map((exercise, index) => {
            if(index === exerciseIndex) {
                setUnsavedChanges(true);
                return { ...exercise, sets };
            }

            return exercise;
        }));
    }
    
    const updateReps = (exerciseIndex: number, reps: number) => {
        setExercises(exercises.map((exercise, index) => {
            if(index === exerciseIndex) {
                setUnsavedChanges(true);
                return { ...exercise, reps };
            }

            return exercise;
        }));
    } 

    const updateWorkout = () => {
        if(workoutExists.current){
            WorkoutService.updateWorkout({
                requestBody:{
                    date: workoutDate,
                    exercises
                }
            })
            .then(() => setSaveSuccess(true))
            .catch((error: ApiError) => {setSaveFailureMessage(
                error.statusText === undefined ? 
                error.toString() : error.statusText
            )});
        } else {
            WorkoutService.addWorkout({
                requestBody: {
                    date: workoutDate,
                    exercises
                }
            })
            .then(() => {
                setSaveSuccess(true);
                setUnsavedChanges(false);
                workoutExists.current = true;
            })
            .catch((error: ApiError) => {setSaveFailureMessage(
                error.statusText === undefined ? 
                error.toString() : error.statusText
            )});
        }
    }

    return (
        <Panel>
            <PanelHeader>
                <Title headingLevel='h2'>{ WORKOUTS }</Title>
            </PanelHeader>
            <Divider />
            <PanelMain>
                <PanelMainBody>
                    <Form>
                        <FormGroup label="Workout Date">
                            {
                                unsavedChanges && (
                                    <Alert title="There are unsaved changes. Click 'Save Workout' to save." variant='warning'/>
                                )
                            }
                            <DatePicker
                                onChange={(_event, str) => setWorkoutDate(str)}
                                value={workoutDate}
                            />
                        </FormGroup>
                        <FormGroup label="Exercises">
                            <Table>
                                <Thead>
                                    <Tr>
                                        <Th>Name</Th>
                                        <Th>Lbs</Th>
                                        <Th>Reps</Th>
                                        <Th>Sets</Th>
                                        <Th>Action</Th>
                                    </Tr>
                                </Thead>
                                <Tbody>
                                    {
                                        exercises.map((exercise: Exercise, index) => (
                                            <Tr key={index}>
                                                <Td>
                                                    <TextInput id={exercise.name} 
                                                        value={exercise.name} 
                                                        onChange={(_event, name: string) => updateName(index, name)}
                                                        aria-label={exercise.name.length > 0 ? exercise.name : "Newly added workout"}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.lbs} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateLbs(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() =>  updateLbs(index, exercise.lbs - WEIGHT_STEP >= 0 ? exercise.lbs - WEIGHT_STEP: 0)}
                                                        onPlus={() => updateLbs(index, exercise.lbs + WEIGHT_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.reps} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateReps(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() => updateReps(index, exercise.reps - REP_STEP >= 0 ? exercise.reps - REP_STEP: 0)}
                                                        onPlus={() => updateReps(index, exercise.reps + REP_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.sets} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateSets(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() => updateSets(index, exercise.sets - SET_STEP >= 0 ? exercise.sets - SET_STEP: 0)}
                                                        onPlus={() => updateSets(index, exercise.sets + SET_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <Button 
                                                        variant={ButtonVariant.danger}
                                                        onClick={() => removeExercise(index)}
                                                    >
                                                        Delete
                                                    </Button>
                                                </Td>
                                            </Tr>
                                        ))
                                    }
                                    <Tr>
                                        <Td colSpan={4}></Td>
                                        <Td>
                                        <Button variant={ButtonVariant.primary} onClick={() => newExercise()}>
                                            Add
                                        </Button>
                                        </Td>
                                    </Tr>
                                </Tbody>
                            </Table>
                        </FormGroup>
                        {
                            saveSuccess && 
                            <Alert
                                title="Workout Saved"
                                variant="success"
                                timeout={10000}
                                onTimeout={() => setSaveSuccess(false)}
                            />
                        }
                        {
                            saveFailureMessage.length !== 0 && 
                            <Alert
                                isInline
                                isExpandable
                                title="Workout failed to save"
                                variant='danger'
                                timeout={10000}
                                onTimeout={() => setSaveFailureMessage('')}
                            >
                                { saveFailureMessage }
                            </Alert>
                        }
                        <ActionGroup>
                            <Button 
                                variant={ButtonVariant.primary}
                                onClick={() => updateWorkout()}
                            >
                                Save Workout
                            </Button>
                        </ActionGroup>
                    </Form>
                </PanelMainBody>
            </PanelMain>
        </Panel>
    )
}
