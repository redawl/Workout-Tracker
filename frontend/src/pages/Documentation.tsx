import { DOCUMENTATION } from "../components/constants"

export const Documentation = () => {
    return (
        <h1>Hello {DOCUMENTATION}!</h1>
    )
}