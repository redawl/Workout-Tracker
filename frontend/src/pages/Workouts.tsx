import { 
    Title, 
    Panel, 
    PanelHeader, 
    Divider, 
    PanelMain, 
    PanelMainBody, 
    DatePicker, 
    Form, 
    FormGroup,
    ActionGroup,
    Button,
    ButtonVariant,
    NumberInput,
    Alert
} from '@patternfly/react-core';
import { REP_STEP, SET_STEP, WEIGHT_STEP, WORKOUTS } from '../components/constants';
import React from 'react';
import { formatDate } from '../helpers/DateHelpers';
import {     
    Thead,
    Tr,
    Th,
    Table,
    Tbody,
    Td
} from '@patternfly/react-table'

import { ApiError, Exercise, ExerciseService, Workout, WorkoutService } from '../workout-api-client';
import { SearchAutocomplete } from '../components/SearchAutocomplete';

export const Workouts = () => {
    const [workoutDate, setWorkoutDate] = React.useState(formatDate(new Date()));
    const [exercises, setExercises] = React.useState([] as Array<Exercise>);
    const [saveSuccess, setSaveSuccess] = React.useState(false);
    const [saveFailureMessage, setSaveFailureMessage] = React.useState('');
    const [unsavedChanges, setUnsavedChanges] = React.useState(false);
    const [exerciseReferences, setExerciseReferences] = React.useState([] as Array<Exercise>);

    const workoutExists = React.useRef(false);

    React.useEffect(() => {
        ExerciseService.getExercises().then((references) => {
            setExerciseReferences([...references]);
        }).catch((apiError: ApiError) => {
            console.error(apiError);
        });
        WorkoutService.getWorkoutByDate({ date: workoutDate })
            .then((workout: Workout) => {
                    workoutExists.current = true;
                setExercises(workout.exercises as Array<Exercise>);
                setUnsavedChanges(false);
            })
            .catch((apiError: ApiError) => {
                if(apiError.status === 404){
                    workoutExists.current = false;
                    setExercises([]);
                } else {
                    console.error(apiError);
                }
            });
    }, [workoutDate]);

    const newExercise = () => {
        setExercises([...exercises,
            {
                name: "",
                lbs: 0,
                sets: 0,
                reps: 0
            }
        ])
        setUnsavedChanges(true);
    }

    const removeExercise = (exerciseIndex: number) => {
        setExercises(exercises.filter((_: Exercise, index: number) => index !== exerciseIndex));
    }

    const updateName = (exerciseIndex: number, exerciseName: string) => {
        setExercises(exercises.map(
            (exercise: Exercise, index: number) => index === exerciseIndex ? { ...exercise, name: exerciseName } : exercise
        ));
        setUnsavedChanges(true);
    }

    const updateLbs = (exerciseIndex: number, lbs: number) => {
        setExercises(exercises.map(
            (exercise: Exercise, index: number) => index === exerciseIndex ? { ...exercise, lbs } : exercise
        ));
        setUnsavedChanges(true);
    } 

    const updateSets = (exerciseIndex: number, sets: number) => {
        setExercises(exercises.map(
            (exercise: Exercise, index: number) => index === exerciseIndex ? { ...exercise, sets } : exercise
        ));
        setUnsavedChanges(true);
    }
    
    const updateReps = (exerciseIndex: number, reps: number) => {
        setExercises(exercises.map((exercise: Exercise, index: number) => {
            if(index === exerciseIndex) {
                setUnsavedChanges(true);
                return { ...exercise, reps };
            }

            return exercise;
        }));
    } 

    const updateWorkout = () => {
        if(workoutExists.current){
            WorkoutService.updateWorkout({
                requestBody:{
                    date: workoutDate,
                    exercises
                }
            })
            .then(() => {
                setSaveSuccess(true);
                setUnsavedChanges(false);
            })
            .catch((error: ApiError) => {setSaveFailureMessage(
                error.statusText === undefined ? 
                error.toString() : error.statusText
            )});
        } else {
            WorkoutService.addWorkout({
                requestBody: {
                    date: workoutDate,
                    exercises
                }
            })
            .then(() => {
                setSaveSuccess(true);
                setUnsavedChanges(false);
                workoutExists.current = true;
            })
            .catch((error: ApiError) => {setSaveFailureMessage(
                error.statusText === undefined ? 
                error.toString() : error.statusText
            )});
        }
    }

    const updateExercise = (exercise: Exercise, index: number) => {
        setExercises(exercises.map((currExercise: Exercise, exerciseIndex: number) => {
            if(exerciseIndex === index){
                return {...exercise};
            }
            return currExercise;
        }));
    }

    return (
        <Panel>
            <PanelHeader>
                <Title headingLevel='h2'>{ WORKOUTS }</Title>
            </PanelHeader>
            <Divider />
            <PanelMain>
                <PanelMainBody>
                    <Form>
                        {
                            unsavedChanges && (
                                <FormGroup>
                                    <Alert title="There are unsaved changes. Click 'Save Workout' to save." variant='warning'/>
                                </FormGroup>
                            )
                        }
                        {
                            saveSuccess && 
                            <FormGroup>
                                <Alert
                                    title="Workout Saved"
                                    variant="success"
                                    timeout={10000}
                                    onTimeout={() => setSaveSuccess(false)}
                                />
                            </FormGroup>
                        }
                        <FormGroup label="Workout Date">
                            <DatePicker
                                onChange={(_event, str) => setWorkoutDate(str)}
                                value={workoutDate}
                            />
                        </FormGroup>
                        <FormGroup label="Exercises">
                            <Table>
                                <Thead>
                                    <Tr>
                                        <Th>Name</Th>
                                        <Th>Lbs</Th>
                                        <Th>Reps</Th>
                                        <Th>Sets</Th>
                                        <Th>Action</Th>
                                    </Tr>
                                </Thead>
                                <Tbody>
                                    {
                                        exercises.map((exercise: Exercise, index: number) => (
                                            <Tr key={index}>
                                                <Td>
                                                    <SearchAutocomplete
                                                        exercises={exerciseReferences}
                                                        index={index}
                                                        startingValue={exercise.name}
                                                        parentOnSelect={(exercise: Exercise, index: number) => updateExercise(exercise, index)}
                                                        parentOnChange={(exerciseIndex: number, exerciseName: string) => updateName(exerciseIndex, exerciseName)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.lbs} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateLbs(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() =>  updateLbs(index, exercise.lbs - WEIGHT_STEP >= 0 ? exercise.lbs - WEIGHT_STEP: 0)}
                                                        onPlus={() => updateLbs(index, exercise.lbs + WEIGHT_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.reps} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateReps(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() => updateReps(index, exercise.reps - REP_STEP >= 0 ? exercise.reps - REP_STEP: 0)}
                                                        onPlus={() => updateReps(index, exercise.reps + REP_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <NumberInput 
                                                        value={exercise.sets} 
                                                        onChange={(event: React.FormEvent<HTMLInputElement>) => updateSets(index, +(event.target as HTMLInputElement).value)}
                                                        onMinus={() => updateSets(index, exercise.sets - SET_STEP >= 0 ? exercise.sets - SET_STEP: 0)}
                                                        onPlus={() => updateSets(index, exercise.sets + SET_STEP)}
                                                    />
                                                </Td>
                                                <Td>
                                                    <Button 
                                                        variant={ButtonVariant.danger}
                                                        onClick={() => removeExercise(index)}
                                                    >
                                                        Delete
                                                    </Button>
                                                </Td>
                                            </Tr>
                                        ))
                                    }
                                    <Tr>
                                        <Td colSpan={4}></Td>
                                        <Td>
                                        <Button variant={ButtonVariant.primary} onClick={() => newExercise()}>
                                            Add
                                        </Button>
                                        </Td>
                                    </Tr>
                                </Tbody>
                            </Table>
                        </FormGroup>
                        {
                            saveFailureMessage.length !== 0 && 
                            <Alert
                                isInline
                                isExpandable
                                title="Workout failed to save"
                                variant='danger'
                                timeout={10000}
                                onTimeout={() => setSaveFailureMessage('')}
                            >
                                { saveFailureMessage }
                            </Alert>
                        }
                        <ActionGroup>
                            <Button 
                                variant={ButtonVariant.primary}
                                onClick={() => updateWorkout()}
                            >
                                Save Workout
                            </Button>
                        </ActionGroup>
                    </Form>
                </PanelMainBody>
            </PanelMain>
        </Panel>
    )
}
