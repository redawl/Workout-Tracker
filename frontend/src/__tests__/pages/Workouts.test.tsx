import { render, waitFor } from '@testing-library/react';
import { Workouts } from '../../pages/Workouts';

jest.mock('../../workout-api-client', () => {
    return {
        ApiError: '',
        Exercise: '',
        ExerciseService: {
            getExercises: jest.fn(() => Promise.resolve([])),
        },
        Workout: '',
        WorkoutService: {
            getWorkoutByDate: jest.fn(() => Promise.reject({
                status: 404
            })),
        },
    };
})

describe('Workouts', () => {
    it('Matches snapshot', async () => {
        const { container } = render(<Workouts />);

        await waitFor(() => expect(container).toMatchSnapshot());
    })
})