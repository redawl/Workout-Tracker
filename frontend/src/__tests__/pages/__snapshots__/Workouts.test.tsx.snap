// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workouts Matches snapshot 1`] = `
<div>
  <div
    class="pf-v5-c-panel"
  >
    <div
      class="pf-v5-c-panel__header"
    >
      <h2
        class="pf-v5-c-title pf-m-xl"
        data-ouia-component-id="OUIA-Generated-Title-1"
        data-ouia-component-type="PF5/Title"
        data-ouia-safe="true"
      >
        Workouts
      </h2>
    </div>
    <hr
      class="pf-v5-c-divider"
    />
    <div
      class="pf-v5-c-panel__main"
    >
      <div
        class="pf-v5-c-panel__main-body"
      >
        <form
          class="pf-v5-c-form"
          novalidate=""
        >
          <div
            class="pf-v5-c-form__group"
          >
            <div
              class="pf-v5-c-form__group-label"
            >
              <label
                class="pf-v5-c-form__label"
              >
                <span
                  class="pf-v5-c-form__label-text"
                >
                  Workout Date
                </span>
              </label>
               
            </div>
            <div
              class="pf-v5-c-form__group-control"
            >
              <div
                class="pf-v5-c-date-picker"
                style="--pf-v5-c-date-picker__input--c-form-control--width-chars: 10;"
              >
                <div
                  class="pf-v5-c-date-picker__input"
                >
                  <div
                    class="pf-v5-c-input-group"
                  >
                    <div
                      class="pf-v5-c-input-group__item"
                    >
                      <span
                        class="pf-v5-c-form-control"
                      >
                        <input
                          aria-invalid="false"
                          aria-label="Date picker"
                          data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                          data-ouia-component-type="PF5/TextInput"
                          data-ouia-safe="true"
                          placeholder="YYYY-MM-DD"
                          type="text"
                          value="2024-12-28"
                        />
                      </span>
                    </div>
                    <div
                      class="pf-v5-c-input-group__item"
                    >
                      <button
                        aria-label="Toggle date picker"
                        class="pf-v5-c-button pf-m-control"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 448 512"
                          width="1em"
                        >
                          <path
                            d="M148 288h-40c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12zm108-12v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 96v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm-96 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm192 0v-40c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12zm96-260v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h48V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h128V12c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h48c26.5 0 48 21.5 48 48zm-48 346V160H48v298c0 3.3 2.7 6 6 6h340c3.3 0 6-2.7 6-6z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-c-form__group"
          >
            <div
              class="pf-v5-c-form__group-label"
            >
              <label
                class="pf-v5-c-form__label"
              >
                <span
                  class="pf-v5-c-form__label-text"
                >
                  Exercises
                </span>
              </label>
               
            </div>
            <div
              class="pf-v5-c-form__group-control"
            >
              <table
                class="pf-v5-c-table pf-m-grid-md"
                data-ouia-component-id="OUIA-Generated-Table-1"
                data-ouia-component-type="PF5/Table"
                data-ouia-safe="true"
                role="grid"
              >
                <thead
                  class="pf-v5-c-table__thead"
                >
                  <tr
                    class="pf-v5-c-table__tr"
                    data-ouia-component-id="OUIA-Generated-TableRow-1"
                    data-ouia-component-type="PF5/TableRow"
                    data-ouia-safe="true"
                  >
                    <th
                      class="pf-v5-c-table__th"
                      scope="col"
                      tabindex="-1"
                    >
                      Name
                    </th>
                    <th
                      class="pf-v5-c-table__th"
                      scope="col"
                      tabindex="-1"
                    >
                      Lbs
                    </th>
                    <th
                      class="pf-v5-c-table__th"
                      scope="col"
                      tabindex="-1"
                    >
                      Reps
                    </th>
                    <th
                      class="pf-v5-c-table__th"
                      scope="col"
                      tabindex="-1"
                    >
                      Sets
                    </th>
                    <th
                      class="pf-v5-c-table__th"
                      scope="col"
                      tabindex="-1"
                    >
                      Action
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="pf-v5-c-table__tbody"
                  role="rowgroup"
                >
                  <tr
                    class="pf-v5-c-table__tr"
                    data-ouia-component-id="OUIA-Generated-TableRow-2"
                    data-ouia-component-type="PF5/TableRow"
                    data-ouia-safe="true"
                  >
                    <td
                      class="pf-v5-c-table__td"
                      colspan="4"
                      tabindex="-1"
                    />
                    <td
                      class="pf-v5-c-table__td"
                      tabindex="-1"
                    >
                      <button
                        aria-disabled="false"
                        class="pf-v5-c-button pf-m-primary"
                        data-ouia-component-id="OUIA-Generated-Button-primary-1"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe="true"
                        type="button"
                      >
                        Add
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            class="pf-v5-c-form__group pf-m-action"
          >
            <div
              class="pf-v5-c-form__group-control"
            >
              <div
                class="pf-v5-c-form__actions"
              >
                <button
                  aria-disabled="false"
                  class="pf-v5-c-button pf-m-primary"
                  data-ouia-component-id="OUIA-Generated-Button-primary-2"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  type="button"
                >
                  Save Workout
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
